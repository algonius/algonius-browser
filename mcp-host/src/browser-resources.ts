export interface BrowserState {
  activeTab?: {
    id?: number;
    url?: string;
    title?: string;
    domState?: any;
  };
  tabs?: Array<{
    id?: number;
    url?: string;
    title?: string;
    active?: boolean;
  }>;
}

export class BrowserResources {
  private browserState: BrowserState = {};

  public updateState(state: BrowserState) {
    this.browserState = state;
    console.error('Browser state updated:', JSON.stringify(state, null, 2).substring(0, 200) + '...');
  }

  public listResources() {
    return [
      {
        uri: 'browser://current/state',
        name: 'Current Browser State',
        mimeType: 'application/json',
        description: 'Complete state of the current active page and all tabs',
      },
      {
        uri: 'browser://current/dom',
        name: 'Current Page DOM',
        mimeType: 'application/json',
        description: 'DOM structure of the current page',
      },
    ];
  }

  public async readResource(uri: string) {
    if (!this.browserState) {
      throw new Error('Browser state not available');
    }

    switch (uri) {
      case 'browser://current/state':
        return {
          contents: [
            {
              uri,
              mimeType: 'application/json',
              text: JSON.stringify(this.browserState, null, 2),
            },
          ],
        };

      case 'browser://current/dom':
        if (!this.browserState.activeTab?.domState) {
          throw new Error('DOM state not available');
        }

        return {
          contents: [
            {
              uri,
              mimeType: 'application/json',
              text: JSON.stringify(this.browserState.activeTab.domState, null, 2),
            },
          ],
        };

      default:
        throw new Error(`Resource not found: ${uri}`);
    }
  }
}
