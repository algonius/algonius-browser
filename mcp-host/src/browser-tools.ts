export interface ActionResult {
  success: boolean;
  message?: string;
  data?: any;
}

export interface ActionCallback {
  (action: string, params: any): Promise<ActionResult>;
}

export class BrowserTools {
  private browserActionCallback: ActionCallback | null = null;

  constructor() {}

  public setBrowserActionCallback(callback: ActionCallback) {
    this.browserActionCallback = callback;
  }

  public listTools() {
    return [
      {
        name: 'navigate_to',
        description: 'Navigate to a specified URL',
        inputSchema: {
          type: 'object',
          properties: {
            url: {
              type: 'string',
              description: 'The URL to navigate to',
            },
          },
          required: ['url'],
        },
      },
      {
        name: 'click_element',
        description: 'Click on a specified element',
        inputSchema: {
          type: 'object',
          properties: {
            selector: {
              type: 'string',
              description: 'CSS selector of the element to click',
            },
          },
          required: ['selector'],
        },
      },
      {
        name: 'add',
        description: 'Add two numbers together for testing purposes',
        inputSchema: {
          type: 'object',
          properties: {
            a: {
              type: 'number',
              description: 'First number to add',
            },
            b: {
              type: 'number',
              description: 'Second number to add',
            },
          },
          required: ['a', 'b'],
        },
      },
    ];
  }

  public async callTool(name: string, args: any) {
    if (!this.browserActionCallback) {
      throw new Error('Browser action callback not set');
    }

    switch (name) {
      case 'navigate_to':
        if (!args.url) {
          throw new Error('URL is required for navigation');
        }
        return this.browserActionCallback('navigate', { url: args.url });

      case 'click_element':
        if (!args.selector) {
          throw new Error('Element selector is required for clicking');
        }
        return this.browserActionCallback('click', { selector: args.selector });

      default:
        throw new Error(`Unknown tool: ${name}`);
    }
  }
}
