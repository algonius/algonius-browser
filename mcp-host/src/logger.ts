/**
 * Logger module for MCP Host
 *
 * Ensures all logs go to stderr to avoid interfering with the Native Messaging protocol
 * which uses stdout for communication with the Chrome extension.
 */

// Log levels in order of verbosity
export enum LogLevel {
  ERROR = 0,
  WARN = 1,
  INFO = 2,
  DEBUG = 3,
}

// Get log level from environment variable or default to INFO
function getLogLevel(): LogLevel {
  const envLevel = process.env.LOG_LEVEL?.toUpperCase();

  if (!envLevel) {
    return LogLevel.INFO; // Default level
  }

  switch (envLevel) {
    case 'ERROR':
      return LogLevel.ERROR;
    case 'WARN':
      return LogLevel.WARN;
    case 'INFO':
      return LogLevel.INFO;
    case 'DEBUG':
      return LogLevel.DEBUG;
    default:
      return LogLevel.INFO;
  }
}

export class Logger {
  private moduleName: string;
  private static logLevel: LogLevel = getLogLevel();

  constructor(moduleName: string) {
    this.moduleName = moduleName;
  }

  /**
   * Format a log message with timestamp, level, module name and params
   */
  private formatMessage(level: string, message: string, args: any[]): string {
    const timestamp = new Date().toISOString();
    const prefix = `[${timestamp}] [${level}] [${this.moduleName}]`;

    let formattedMessage = `${prefix} ${message}`;

    // If there are additional arguments, stringify and append them
    if (args.length > 0) {
      // Format each argument
      const formattedArgs = args
        .map(arg => {
          if (typeof arg === 'object') {
            try {
              return JSON.stringify(arg);
            } catch (e) {
              return '[Circular Object]';
            }
          }
          return String(arg);
        })
        .join(' ');

      formattedMessage += ` ${formattedArgs}`;
    }

    return formattedMessage;
  }

  /**
   * Log an error message (always logged)
   */
  error(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.ERROR) {
      const formattedMessage = this.formatMessage('ERROR', message, args);
      console.error(formattedMessage);
    }
  }

  /**
   * Log a warning message
   */
  warn(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.WARN) {
      const formattedMessage = this.formatMessage('WARN', message, args);
      console.error(formattedMessage); // Still using stderr
    }
  }

  /**
   * Log an info message
   */
  info(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.INFO) {
      const formattedMessage = this.formatMessage('INFO', message, args);
      console.error(formattedMessage); // Still using stderr
    }
  }

  /**
   * Log a debug message
   */
  debug(message: string, ...args: any[]): void {
    if (Logger.logLevel >= LogLevel.DEBUG) {
      const formattedMessage = this.formatMessage('DEBUG', message, args);
      console.error(formattedMessage); // Still using stderr
    }
  }

  /**
   * Set log level globally for all loggers
   */
  static setLogLevel(level: LogLevel): void {
    Logger.logLevel = level;
  }

  /**
   * Get current log level
   */
  static getLogLevel(): LogLevel {
    return Logger.logLevel;
  }
}

/**
 * Create a new logger instance
 * @param moduleName The module name for this logger
 * @returns A logger instance
 */
export function createLogger(moduleName: string): Logger {
  return new Logger(moduleName);
}
