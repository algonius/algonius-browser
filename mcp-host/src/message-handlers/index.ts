/**
 * Message handlers for MCP Host
 *
 * This file exports all message handlers from their individual files
 */

export * from './types.js';
export * from './status-handler.js';
export * from './ping-handler.js';
export * from './shutdown-handler.js';
export * from './mcp-server-stop-handler.js';
export * from './mcp-server-status-handler.js';
