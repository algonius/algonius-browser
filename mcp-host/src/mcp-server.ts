import { McpServer as SdkMcpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StreamableHTTPServerTransport } from '@modelcontextprotocol/sdk/server/streamableHttp.js';
import express from 'express';
import { randomUUID } from 'crypto';
import { isInitializeRequest } from '@modelcontextprotocol/sdk/types.js';
import { createLogger, LogLevel, Logger } from './logger.js';
import { BrowserResources } from './browser-resources.js';
import { BrowserTools, ActionCallback } from './browser-tools.js';

// Implementation of McpServer class used in tests
export class McpServer {
  private browserResources: BrowserResources;
  private browserTools: BrowserTools;
  private actionCallback: ActionCallback | null = null;
  private logger;
  private transport: StreamableHTTPServerTransport | null = null;
  private name: string;
  private version: string;

  constructor(options?: { name?: string; version?: string }) {
    this.browserResources = new BrowserResources();
    this.browserTools = new BrowserTools();
    this.logger = createLogger('mcp-server');
    this.name = options?.name || 'nanobrowser-mcp';
    this.version = options?.version || '1.0.0';
    this.logger.info(`McpServer instance created (${this.name} v${this.version})`);
  }

  /**
   * Connects the MCP server to a transport
   * @param transport The transport to connect to
   */
  public async connect(transport: StreamableHTTPServerTransport): Promise<void> {
    this.transport = transport;
    this.logger.info('Connected to transport');
  }

  /**
   * Registers an action callback for handling browser operations
   * @param callback The callback function for executing browser actions
   */
  public registerActionCallback(callback: ActionCallback): void {
    this.actionCallback = callback;
    this.browserTools.setBrowserActionCallback(callback);
    this.logger.info('Action callback registered');
  }

  /**
   * Sets the browser state
   * @param state The browser state to set
   */
  public setBrowserState(state: any): void {
    this.browserResources.updateState(state);
    this.logger.debug('Browser state updated');
  }

  /**
   * Lists all available resources
   * @returns A promise that resolves to the list of resources
   */
  public async handleListResources(): Promise<{ resources: any[] }> {
    return {
      resources: this.browserResources.listResources(),
    };
  }

  /**
   * Reads a resource by URI
   * @param uri The URI of the resource to read
   * @returns A promise that resolves to the resource content
   */
  public async handleReadResource(uri: string): Promise<any> {
    try {
      return await this.browserResources.readResource(uri);
    } catch (error) {
      this.logger.error(`Error reading resource ${uri}:`, error);
      throw error;
    }
  }

  /**
   * Lists all available tools
   * @returns A promise that resolves to the list of tools
   */
  public async handleListTools(): Promise<{ tools: any[] }> {
    return {
      tools: this.browserTools.listTools(),
    };
  }

  /**
   * Calls a tool with arguments
   * @param name The name of the tool to call
   * @param args The arguments to pass to the tool
   * @returns A promise that resolves to the tool result
   */
  public async handleCallTool(name: string, args: any): Promise<any> {
    try {
      this.logger.debug(`Executing tool: ${name} with args:`, args);
      return await this.browserTools.callTool(name, args);
    } catch (error) {
      this.logger.error(`Error executing tool ${name}:`, error);
      throw error;
    }
  }
}

// Extend the SDK's McpServer type for our custom handlers
declare module '@modelcontextprotocol/sdk/server/mcp.js' {
  interface McpServer {
    handleListResources?: () => Promise<{ resources: any[] }>;
    handleReadResource?: (uri: string) => Promise<any>;
    handleListTools?: () => Promise<{ tools: any[] }>;
    handleCallTool?: (name: string, args: any) => Promise<any>;
  }
}

/**
 * Configuration for the MCP server
 */
export interface McpServerConfig {
  port: number;
  logLevel: string;
}

/**
 * Manages the MCP server and HTTP transport
 */
export class McpServerManager {
  private logger;
  private httpServer: any = null;
  private mcpServer: McpServer;
  private config: McpServerConfig;
  private browserResources: BrowserResources;
  private browserTools: BrowserTools;
  private isRunning: boolean = false;

  /**
   * Creates a new MCP server manager
   * @param config The server configuration
   */
  constructor(config: McpServerConfig) {
    this.config = config;
    this.logger = createLogger('mcp-server');

    // Set the log level based on the config
    this.setLogLevel(config.logLevel);

    // Initialize browser resources and tools
    this.browserResources = new BrowserResources();
    this.browserTools = new BrowserTools();

    // Create MCP server instance
    this.mcpServer = new McpServer({
      name: 'nanobrowser-mcp',
      version: '1.0.0',
    });

    this.logger.info('MCP server manager initialized');
  }

  /**
   * Sets the browser action callback to handle browser operations
   * @param callback The callback function to execute browser actions
   */
  public setBrowserActionCallback(callback: ActionCallback) {
    this.mcpServer.registerActionCallback(callback);
    this.browserTools.setBrowserActionCallback(callback);
    this.logger.info('Browser action callback set');
  }

  /**
   * Updates the browser state
   * @param state The current browser state
   */
  public setBrowserState(state: any) {
    this.mcpServer.setBrowserState(state);
    this.browserResources.updateState(state);
    this.logger.debug('Browser state updated');
  }

  /**
   * Starts the MCP server
   * @returns A promise that resolves when the server is started
   */
  public async start(): Promise<boolean> {
    if (this.isRunning) {
      this.logger.warn('MCP server is already running');
      return false;
    }

    try {
      // Register browser resources and tools with the MCP server
      this.registerResourcesAndTools();

      // Start the HTTP server
      await this.startHttpServer();

      this.isRunning = true;
      this.logger.info(`MCP server started on port ${this.config.port}`);
      return true;
    } catch (error) {
      this.logger.error('Failed to start MCP server:', error);
      return false;
    }
  }

  /**
   * Registers browser resources and tools with the MCP server
   */
  private registerResourcesAndTools() {
    this.logger.info('Registering browser resources and tools');

    // Register resource handlers with SdkMcpServer
    this.mcpServer.handleListResources = async () => {
      this.logger.debug('Handling listResources request');
      return {
        resources: this.browserResources.listResources(),
      };
    };

    this.mcpServer.handleReadResource = async (uri: string) => {
      this.logger.debug(`Handling readResource request for ${uri}`);
      try {
        return await this.browserResources.readResource(uri);
      } catch (error) {
        this.logger.error(`Error reading resource ${uri}:`, error);
        throw error;
      }
    };

    // Register tool handlers with SdkMcpServer
    this.mcpServer.handleListTools = async () => {
      this.logger.debug('Handling listTools request');
      return {
        tools: this.browserTools.listTools(),
      };
    };

    this.mcpServer.handleCallTool = async (name: string, args: any) => {
      this.logger.debug(`Handling callTool request for ${name}`);
      try {
        this.logger.debug(`Executing tool: ${name} with args:`, args);
        return await this.browserTools.callTool(name, args);
      } catch (error) {
        this.logger.error(`Error executing tool ${name}:`, error);
        throw error;
      }
    };

    // Log registered resources and tools
    const resources = this.browserResources.listResources();
    this.logger.info(`Made ${resources.length} resources available through MCP server`);
    this.logger.debug('Available resources: ' + resources.map(r => r.uri).join(', '));

    const tools = this.browserTools.listTools();
    this.logger.info(`Made ${tools.length} tools available through MCP server`);
    this.logger.debug('Available tools: ' + tools.map(t => t.name).join(', '));
  }

  /**
   * Starts the HTTP server with StreamableHTTPServerTransport
   */
  private async startHttpServer() {
    // Create Express application
    const app = express();

    // Initialize transports map to store session connections
    const transports: { [sessionId: string]: StreamableHTTPServerTransport } = {};

    // Set up HTTP routes for MCP communication
    app.use(express.json());

    // Handle POST requests (client to server communication)
    app.post('/mcp', async (req, res) => {
      try {
        // Check if there's an existing session ID
        const sessionId = req.headers['mcp-session-id'] as string | undefined;
        let transport: StreamableHTTPServerTransport;

        if (sessionId && transports[sessionId]) {
          // Reuse existing transport
          transport = transports[sessionId];
        } else if (!sessionId && isInitializeRequest(req.body)) {
          // New initialization request
          this.logger.info('Creating new MCP session');

          // Create new transport with session ID generator
          transport = new StreamableHTTPServerTransport({
            sessionIdGenerator: () => randomUUID(),
            onsessioninitialized: sessionId => {
              // Store transport instance by session ID
              transports[sessionId] = transport;
              this.logger.debug(`Session initialized: ${sessionId}`);
            },
          });

          // Clean up transport when closed
          transport.onclose = () => {
            if (transport.sessionId) {
              delete transports[transport.sessionId];
              this.logger.debug(`Session closed: ${transport.sessionId}`);
            }
          };

          // Connect MCP server to the transport
          await this.mcpServer.connect(transport);
        } else {
          // Invalid request
          this.logger.warn('Invalid request: No valid session ID');
          res.status(400).json({
            jsonrpc: '2.0',
            error: {
              code: -32000,
              message: 'Invalid request: No valid session ID provided',
            },
            id: null,
          });
          return;
        }

        // Handle the request
        await transport.handleRequest(req, res, req.body);
      } catch (error) {
        this.logger.error('Error handling POST request:', error);
        res.status(500).json({
          jsonrpc: '2.0',
          error: {
            code: -32000,
            message: error instanceof Error ? error.message : 'Unknown error',
          },
          id: null,
        });
      }
    });

    // Generic session request handler for GET and DELETE
    const handleSessionRequest = async (req: express.Request, res: express.Response) => {
      try {
        const sessionId = req.headers['mcp-session-id'] as string | undefined;
        if (!sessionId || !transports[sessionId]) {
          this.logger.warn(`Invalid or missing session ID: ${sessionId}`);
          res.status(400).send('Invalid or missing session ID');
          return;
        }

        const transport = transports[sessionId];
        await transport.handleRequest(req, res);
      } catch (error) {
        this.logger.error('Error handling session request:', error);
        res.status(500).send('Internal server error');
      }
    };

    // Handle GET requests (server to client notifications, using SSE)
    app.get('/mcp', handleSessionRequest);

    // Handle DELETE requests (session termination)
    app.delete('/mcp', handleSessionRequest);

    // Start the HTTP server
    return new Promise<void>((resolve, reject) => {
      try {
        this.httpServer = app.listen(this.config.port, '127.0.0.1', () => {
          this.logger.info(`HTTP server listening on http://localhost:${this.config.port}`);
          resolve();
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Shuts down the MCP server
   * @returns A promise that resolves when the server is shut down
   */
  public async shutdown(): Promise<boolean> {
    if (!this.isRunning) {
      this.logger.warn('MCP server is not running');
      return false;
    }

    try {
      // Close the HTTP server
      if (this.httpServer) {
        await new Promise<void>(resolve => {
          this.httpServer.close(() => {
            this.logger.info('HTTP server closed');
            resolve();
          });
        });
        this.httpServer = null;
      }

      this.isRunning = false;
      this.logger.info('MCP server shutdown complete');
      return true;
    } catch (error) {
      this.logger.error('Error shutting down MCP server:', error);
      return false;
    }
  }

  /**
   * Gets the running status of the MCP server
   * @returns True if the server is running, false otherwise
   */
  public isServerRunning(): boolean {
    return this.isRunning;
  }

  /**
   * Sets the log level for the MCP server
   * @param level The log level to set
   */
  private setLogLevel(level: string): void {
    switch (level.toUpperCase()) {
      case 'ERROR':
        Logger.setLogLevel(LogLevel.ERROR);
        break;
      case 'WARN':
        Logger.setLogLevel(LogLevel.WARN);
        break;
      case 'INFO':
        Logger.setLogLevel(LogLevel.INFO);
        break;
      case 'DEBUG':
        Logger.setLogLevel(LogLevel.DEBUG);
        break;
      default:
        Logger.setLogLevel(LogLevel.INFO);
        break;
    }
    this.logger.info(`Log level set to ${level.toUpperCase()}`);
  }
}
