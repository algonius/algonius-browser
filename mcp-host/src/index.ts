import express from 'express';
import { ActionResult } from './browser-tools';
import { BrowserMcpServer } from './http-mcp-server';
import { McpServer } from './mcp-server';
import { NativeMessaging } from './messaging';

// Create Express app for Vite compatibility
const app = express();

// Log startup information
console.error('Starting Nanobrowser MCP Native Messaging Host');
console.error(`Process ID: ${process.pid}`);
console.error(`Node.js version: ${process.version}`);
console.error(`Platform: ${process.platform}`);

// Initialize components
const messaging = new NativeMessaging();
const mcpServer = new McpServer();

// Initialize the HTTP-based MCP server
const httpMcpServer = new BrowserMcpServer();

// Register message handlers from extension to MCP servers
messaging.registerHandler('setBrowserState', async data => {
  if (!data.state) {
    throw new Error('No state provided');
  }

  // Update both MCP servers
  mcpServer.setBrowserState(data.state);
  httpMcpServer.setBrowserState(data.state);

  return { success: true };
});

// Register message handlers for MCP server requests
messaging.registerHandler('listResources', async () => {
  const result = await mcpServer.handleListResources();
  return result;
});

messaging.registerHandler('readResource', async data => {
  if (!data.uri) {
    throw new Error('No URI provided');
  }

  const result = await mcpServer.handleReadResource(data.uri);
  return result;
});

messaging.registerHandler('listTools', async () => {
  const result = await mcpServer.handleListTools();
  return result;
});

messaging.registerHandler('callTool', async data => {
  if (!data.name || !data.arguments) {
    throw new Error('Tool name and arguments are required');
  }

  const result = await mcpServer.handleCallTool(data.name, data.arguments);
  return result;
});

// Shared action callback for both MCP servers to send browser actions back to the extension
const handleAction = async (action: string, params: any): Promise<ActionResult> => {
  try {
    // Send the action request to the extension
    messaging.sendMessage({
      type: 'executeAction',
      action,
      params,
    });

    // This is a simplified approach - in a real implementation we would wait for a response
    return {
      success: true,
      message: `Action ${action} requested`,
      data: { action, params },
    };
  } catch (error) {
    return {
      success: false,
      message: error instanceof Error ? error.message : String(error),
    };
  }
};

// Register action callback for both MCP servers
mcpServer.registerActionCallback(handleAction);
httpMcpServer.registerActionCallback(handleAction);

// Start the HTTP MCP server
httpMcpServer.start().catch(error => {
  console.error('Failed to start HTTP MCP server:', error);
});

// Handle exit signals
process.on('SIGINT', async () => {
  console.error('Received SIGINT signal, exiting...');
  await httpMcpServer.stop().catch(error => {
    console.error('Error stopping HTTP MCP server:', error);
  });
  process.exit(0);
});

process.on('SIGTERM', async () => {
  console.error('Received SIGTERM signal, exiting...');
  await httpMcpServer.stop().catch(error => {
    console.error('Error stopping HTTP MCP server:', error);
  });
  process.exit(0);
});

// Notify that we're ready
console.error('Nanobrowser MCP Native Messaging Host is ready');

// Configure Express routes for Vite compatibility
app.get('/', (req: express.Request, res: express.Response) => {
  res.send('Nanobrowser MCP Host is running');
});

// Add health check endpoint
app.get('/health', (req: express.Request, res: express.Response) => {
  res.json({
    status: 'ok',
    version: '0.1.0',
    services: {
      mcpServer: 'running',
      httpMcpServer: 'running',
    },
  });
});

// Export the Express app for Vite
export default app;
