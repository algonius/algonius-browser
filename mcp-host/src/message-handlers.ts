/**
 * Message handlers for MCP Host
 *
 * This file contains implementations of various message handlers
 * that process incoming messages from the Chrome extension.
 */

import { createLogger } from './logger.js';
import { McpServerManager, McpServerConfig } from './mcp-server.js';

/**
 * Common interface for all message handlers
 */
export interface MessageHandler {
  handle(data: any): Promise<any>;
}

/**
 * Configuration for the status handler
 */
export interface StatusHandlerConfig {
  startTime: number;
  version: string;
  runMode: string;
}

/**
 * Handler for status requests
 * Responds with the current status of the MCP Host
 */
export class StatusHandler implements MessageHandler {
  private logger = createLogger('status-handler');
  private config: StatusHandlerConfig;

  constructor(config: StatusHandlerConfig) {
    this.config = config;
  }

  /**
   * Handles a getStatus message
   * @param data Message payload (not used for status requests)
   * @returns Promise resolving to status data
   */
  async handle(data: any): Promise<any> {
    this.logger.debug('Handling status request');

    return {
      data: {
        isConnected: true,
        startTime: this.config.startTime,
        version: this.config.version,
        runMode: this.config.runMode,
      },
    };
  }
}

/**
 * Type definition for ping callback
 */
export type PingCallback = (timestamp: number) => void;

/**
 * Handler for ping (heartbeat) requests
 * Responds with the current timestamp
 */
export class PingHandler implements MessageHandler {
  private logger = createLogger('ping-handler');
  /**
   * Optional callback to be called when a ping is received
   * This can be used to track the last ping time
   */
  public onPing?: PingCallback;

  /**
   * Handles a ping message
   * @param data Message payload (not used for ping requests)
   * @returns Promise resolving to timestamp data
   */
  async handle(data: any): Promise<any> {
    this.logger.debug('Handling ping request');

    const timestamp = Date.now();

    // Call the callback if it's registered
    if (this.onPing) {
      this.onPing(timestamp);
    }

    return {
      timestamp,
    };
  }
}

/**
 * Type definition for the cleanup function
 */
export type CleanupFunction = () => Promise<void>;

/**
 * Handler for shutdown requests
 * Performs cleanup and gracefully exits the process
 */
export class ShutdownHandler implements MessageHandler {
  private logger = createLogger('shutdown-handler');
  private cleanupFn: CleanupFunction;

  /**
   * @param cleanupFn Function to call for resource cleanup before shutdown
   */
  constructor(cleanupFn: CleanupFunction) {
    this.cleanupFn = cleanupFn;
  }

  /**
   * Handles a shutdown message
   * @param data Message payload (not used for shutdown requests)
   * @returns Promise resolving to success status
   */
  async handle(data: any): Promise<any> {
    this.logger.info('Handling shutdown request');

    try {
      // Perform cleanup
      await this.cleanupFn();

      // Schedule exit after response is sent
      setTimeout(() => {
        this.logger.info('Exiting process after successful shutdown');
        process.exit(0);
      }, 500);

      return { success: true };
    } catch (error) {
      this.logger.error('Error during shutdown cleanup:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Handler for starting the MCP server
 */
export class McpServerStartHandler implements MessageHandler {
  private logger = createLogger('mcp-server-start-handler');
  private mcpServerManager: McpServerManager | null = null;

  /**
   * Creates an instance of the MCP server start handler
   * @param browserActionCallback Optional callback for browser actions
   */
  constructor(private browserActionCallback?: any) {
    this.logger.debug('Initialized MCP server start handler');
  }

  /**
   * Handles a start MCP server message
   * @param data Message payload containing server configuration
   * @returns Promise resolving to success status
   */
  async handle(data: any): Promise<any> {
    this.logger.info('Handling MCP server start request');

    try {
      // Extract configuration from the request
      const config: McpServerConfig = {
        port: data.port || 3000,
        logLevel: data.logLevel || 'info',
      };

      this.logger.info(`Starting MCP server on port ${config.port} with log level ${config.logLevel}`);

      // Create MCP server manager if it doesn't exist
      if (!this.mcpServerManager) {
        this.mcpServerManager = new McpServerManager(config);

        // Set browser action callback if provided
        if (this.browserActionCallback) {
          this.mcpServerManager.setBrowserActionCallback(this.browserActionCallback);
        }
      }

      // Start the server
      const result = await this.mcpServerManager.start();

      return {
        success: result,
        message: result ? 'MCP server started successfully' : 'MCP server already running',
        config,
      };
    } catch (error) {
      this.logger.error('Error starting MCP server:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Gets the MCP server manager instance
   * @returns The MCP server manager
   */
  getMcpServerManager(): McpServerManager | null {
    return this.mcpServerManager;
  }
}

/**
 * Handler for stopping the MCP server
 */
export class McpServerStopHandler implements MessageHandler {
  private logger = createLogger('mcp-server-stop-handler');

  /**
   * Creates an instance of the MCP server stop handler
   * @param getMcpServerManager Function to get the current MCP server manager instance
   */
  constructor(private getMcpServerManager: () => McpServerManager | null) {
    this.logger.debug('Initialized MCP server stop handler');
  }

  /**
   * Handles a stop MCP server message
   * @param data Message payload (not used for stop requests)
   * @returns Promise resolving to success status
   */
  async handle(data: any): Promise<any> {
    this.logger.info('Handling MCP server stop request');

    try {
      const mcpServerManager = this.getMcpServerManager();

      if (!mcpServerManager) {
        this.logger.warn('MCP server manager not found');
        return {
          success: false,
          message: 'MCP server not initialized',
        };
      }

      if (!mcpServerManager.isServerRunning()) {
        this.logger.warn('MCP server not running');
        return {
          success: false,
          message: 'MCP server not running',
        };
      }

      // Stop the server
      const result = await mcpServerManager.shutdown();

      return {
        success: result,
        message: result ? 'MCP server stopped successfully' : 'Failed to stop MCP server',
      };
    } catch (error) {
      this.logger.error('Error stopping MCP server:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Handler for getting the MCP server status
 */
export class McpServerStatusHandler implements MessageHandler {
  private logger = createLogger('mcp-server-status-handler');

  /**
   * Creates an instance of the MCP server status handler
   * @param getMcpServerManager Function to get the current MCP server manager instance
   */
  constructor(private getMcpServerManager: () => McpServerManager | null) {
    this.logger.debug('Initialized MCP server status handler');
  }

  /**
   * Handles a get MCP server status message
   * @param data Message payload (not used for status requests)
   * @returns Promise resolving to status data
   */
  async handle(data: any): Promise<any> {
    this.logger.debug('Handling MCP server status request');

    const mcpServerManager = this.getMcpServerManager();
    const isRunning = mcpServerManager?.isServerRunning() || false;

    return {
      isRunning,
      config: mcpServerManager
        ? {
            port: mcpServerManager['config']?.port,
            logLevel: mcpServerManager['config']?.logLevel,
          }
        : null,
    };
  }
}
