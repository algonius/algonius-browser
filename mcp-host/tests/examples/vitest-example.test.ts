/**
 * Example test file demonstrating Vitest usage with MCP Host
 *
 * This file shows a basic test setup with Vitest and can be used
 * as a template for migrating Jest tests to Vitest.
 */

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { ActionResult } from '../../src/browser-tools';
import { McpServer } from '../../src/mcp-server';

// Mock dependencies
vi.mock('../../src/browser-resources', () => ({
  BrowserResources: vi.fn().mockImplementation(() => ({
    readResource: vi.fn().mockResolvedValue({
      contents: [
        {
          uri: 'browser://state',
          mimeType: 'application/json',
          text: JSON.stringify({ data: 'mock-resource-data' }),
        },
      ],
    }),
  })),
}));

describe('McpServer', () => {
  let mcpServer: McpServer;

  // Setup before each test
  beforeEach(() => {
    // Reset mocks before each test
    vi.clearAllMocks();

    // Create a new instance for each test
    mcpServer = new McpServer();
  });

  // Cleanup after each test
  afterEach(() => {
    vi.resetAllMocks();
  });

  // Test case using async/await
  it('should handle resource requests correctly', async () => {
    // Arrange
    const uri = 'browser://state';

    // Act
    const result = await mcpServer.handleReadResource(uri);

    // Assert
    expect(result).toHaveProperty('contents');
    expect(result.contents[0]).toHaveProperty('text');

    // Parse the JSON text to get the data
    const parsedData = JSON.parse(result.contents[0].text);
    expect(parsedData).toHaveProperty('data');
    expect(parsedData.data).toBe('mock-resource-data');
  });

  // Test case with spy
  it('should register action callbacks correctly', () => {
    // Arrange
    const mockCallback = vi.fn(async (action: string, params: any): Promise<ActionResult> => {
      return { success: true, message: 'Mock action executed' };
    });

    // Act
    mcpServer.registerActionCallback(mockCallback);

    // Assert - we can only test this indirectly since handleCallTool is private
    // In a real test, we would test the effects of registering the callback
    expect(mcpServer).toBeTruthy();
  });

  // Test with mock timer
  it('should handle timeouts correctly', async () => {
    // Setup fake timers
    vi.useFakeTimers();

    // Arrange
    const slowCallback = vi.fn().mockImplementation(() => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve({ success: true });
        }, 5000);
      });
    });

    // Register the slow callback
    mcpServer.registerActionCallback(slowCallback);

    // Act & Assert - in a real test we would test timeout behavior
    // This is just an example of using fake timers
    vi.advanceTimersByTime(5000);

    // Restore real timers
    vi.useRealTimers();
  });

  // Snapshot test example
  it('should match snapshot for tool list', async () => {
    // Act
    const tools = await mcpServer.handleListTools();

    // Assert
    expect(tools).toMatchSnapshot();
  });
});
