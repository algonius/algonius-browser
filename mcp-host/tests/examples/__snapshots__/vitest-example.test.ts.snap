// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`McpServer > should match snapshot for tool list 1`] = `
{
  "tools": [
    {
      "description": "Navigate to a specified URL",
      "inputSchema": {
        "properties": {
          "url": {
            "description": "The URL to navigate to",
            "type": "string",
          },
        },
        "required": [
          "url",
        ],
        "type": "object",
      },
      "name": "navigate_to",
    },
    {
      "description": "Click on a specified element",
      "inputSchema": {
        "properties": {
          "selector": {
            "description": "CSS selector of the element to click",
            "type": "string",
          },
        },
        "required": [
          "selector",
        ],
        "type": "object",
      },
      "name": "click_element",
    },
  ],
}
`;
